src/java/io/bazel/rulesscala/worker/Worker.java

```diff
@@ -0,0 +1,158 @@
+package io.bazel.rulesscala.worker;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.PrintStream;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.lang.SecurityManager;
+import java.security.Permission;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.nio.charset.StandardCharsets;
+
+import com.google.devtools.build.lib.worker.WorkerProtocol;
+
+public final class Worker {
+
+    public static interface Interface {
+	public void work(String[] args) throws Exception;
+    }
+
+    public static void workerMain(String workerArgs[], Interface workerInterface) throws Exception {
+	if (workerArgs.length > 0 && workerArgs[0].equals("--persistent_worker")) {
+	    persistentWorkerMain(workerInterface);
+	} else {
+	    ephemeralWorkerMain(workerArgs, workerInterface);
+	}
+    }
+
+    private static void persistentWorkerMain(Interface workerInterface) throws Exception {
+	System.setSecurityManager(new SecurityManager() {
+		@Override
+		public void checkPermission(Permission permission) {
+		    Matcher matcher = exitPattern.matcher(permission.getName());
+		    if (matcher.find())
+			throw new ExitTrapped(Integer.parseInt(matcher.group(1)));
+		}
+	    });
+
+	InputStream stdin = System.in;
+	PrintStream stdout = System.out;
+	PrintStream stderr = System.err;
+	ByteArrayOutputStream outStream = new SmartByteArrayOutputStream();
+	PrintStream out = new PrintStream(outStream);
+
+	// We can't support stdin, so assign it to read from an empty buffer
+	System.setIn(new ByteArrayInputStream(new byte[0]));
+
+	System.setOut(out);
+	System.setErr(out);
+
+	try {
+	    while (true) {
+		WorkerProtocol.WorkRequest request =
+		    WorkerProtocol.WorkRequest.parseDelimitedFrom(stdin);
+
+		int code = 0;
+
+		try {
+		    workerInterface.work(stringListToArray(request.getArgumentsList()));
+		} catch (ExitTrapped e) {
+		    code = e.code;
+		} catch (Exception e) {
+		    System.err.println(e.getMessage());
+		    e.printStackTrace();
+		    code = 1;
+		}
+
+		WorkerProtocol.WorkResponse.newBuilder()
+		    .setOutput(outStream.toString())
+		    .setExitCode(code)
+		    .build()
+		    .writeDelimitedTo(stdout);
+
+		out.flush();
+		outStream.reset();
+		System.gc();
```

Comment:
Bazel had it in the finally clause of every request, wdyt about moving it to the finally as well?

Reply:
Works for me 👍

Reply:
Looks like I'd need to introduce a new try catch block. Shouldn't be a problem.

Reply:
I refactored things a bit more so this should should be handled now.